### PR TITLE
Update to GHC 9.4

### PR DESCRIPTION
**Pull request by 9999years**
This updates `retrie` for compatibility with GHC 9.4, which is necessary for improved GHC 9.4 support from `haskell-language-server`.

HLS PR based on these changes: https://github.com/haskell/haskell-language-server/pull/3317

### CODE REVIEW

File: Retrie/Expr.hs
Comment (josefs) [inline Retrie/Expr.hs]: Can you please revert this change.

**Review by josefs — CHANGES_REQUESTED** [review]
This is great! Thanks for this effort. 

Three things:
* We currently don't have CI enabled for 9.4. It would make sense to add that as part of this PR.
* Can you please sign the Facebook CLA check so that we can commit this change?
* There seems to be an accidental change in the code that I've pointed out.

File: Retrie/Expr.hs
Comment (wavewave) [reply Retrie/Expr.hs]: oops. reverted it. Thanks for catching that!

**Comment by wavewave** [comment]
Thanks. I am going to work on CI setup now. For CLA, we are currently waiting for the internal process on our side. :-)

**Comment by 9999years** [comment]
OK, the CLA is signed and should be good now.

**Comment by wavewave** [comment]
@josefs : to check with CI, we need an approval for running workflows as you can see in this github message box. Would you or any repo maintainer do that for us? Thank you very much!
Screen Shot 2022-11-10 at 1 33 43 PM

**Comment by josefs** [comment]
Many thanks for your contribution! Merged.

**Comment by 9999years** [comment]
Thanks! Let us know when you've cut a release, it'll be helpful for getting the `haskell-language-server` support in place.

**Comment by pepeiborra** [comment]
> Thanks! Let us know when you've cut a release, it'll be helpful for getting the `haskell-language-server` support in place.

retire 1.2.1 is now in Hackage

**Pull request merged**
